### PR TITLE
it8xxx2/shi: wrap pm policy state get/put

### PR DESCRIPTION
**Pull request by Dino-Li**
When CONFIG_ASSERT is enabled, assertion for unbalanced state lock get/put is triggered. This commit leverages shi drivers from other SoCs to resolve the assertion.

### CODE REVIEW

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c [hunk 1]
```diff
@@ -63,23 +63,67 @@ struct shi_it8xxx2_cfg {
...
+	if (atomic_test_and_clear_bit(data->pm_policy_state_flag, flag) == 1) {
+		pm_policy_state_lock_put(PM_STATE_STANDBY, PM_ALL_SUBSTATES);
+		k_work_cancel_delayable(&data->cs_off_put);
+	}
+}
+
+/*
+ * When AP is In S0 state, AP assert CS of SPI for a new command transaction.
+ * Under the condition, SoC will not enter deep sleep until AP de-assert the CS.
+ * But If AP is powered off (G3 state) CS will go low (assert CS).
```
Comment (niedzwiecki-dawid) [inline subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: Am I understanding correctly, that CS always goes low (asserted) if AP enters G3?

**Review by niedzwiecki-dawid — COMMENTED** [review]
Nice catch!

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 1 above)
Comment (Dino-Li) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: Right.

Comment (niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: So there is only internal PU in AP, which is disconnected when AP enters G3? Despite this change, I'm wondering how it is possible that CS goes low, shouldn't PU be somewhere?

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c [hunk 2]
```diff
@@ -63,23 +63,67 @@ struct shi_it8xxx2_cfg {
...
+ * is not for starting a new command transaction.
+ * This handler is used to resolve the situation.
+ */
+static void cs_off_put_handler(struct k_work *item)
+{
+	struct k_work_delayable *dwork = k_work_delayable_from_work(item);
+	struct shi_it8xxx2_data *data = CONTAINER_OF(dwork, struct shi_it8xxx2_data, cs_off_put);
+	int key;
+
+	key = irq_lock();
```
Comment (niedzwiecki-dawid) [inline subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:122]: Why do we need irq_lock here?

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 1 above)
Comment (Dino-Li) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: If I understand correctly, no external PU on AP_SPI_EC_CS_L.
And AP's power rail will be cut off in G3 state, so internal PU will not work.

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 2 above)
Comment (Dino-Li) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:122]: The above pm policy get/put is called in ISR, and this put is called in workqueue thread.
I want to make sure there is no potential race condition between them.

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 1 above)
Comment (niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: Is there any possibility to have internal PU on the EC side? I've been working with stm32 (SPI) and they do it [this way](https://github.com/zephyrproject-rtos/hal_stm32/blob/main/dts/st/h7/stm32h743vitx-pinctrl.dtsi#L2079). Additionally, I can see that ITE also has the pull-up for some pins with an alternate function, e.g. [keyboard](https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/riscv/ite/it8xxx2-pinctrl-map.dtsi#L115).

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 2 above)
Comment (niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:122]: I don't it is necessary. Please correct me if I'm wrong. The handler can be called only if CS is asserted (de-assertion cancels k_work). So we stop irq only low->high, which will do the same anyway. So I think it is not needed, especially with the atomic-test guard.

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 1 above)
Comment (Dino-Li) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: On IT8xxx2, enable internal PU on EC's CS will pull this pin up to PP1800_EC_VCC connected to PP1800_VIO18_S3. PP1800_VIO18_S3 is turned off in G3 state. (Please see the attachment)

ec-vcc-1p8v

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 2 above)
Comment (Dino-Li) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c]: Removed the irq_lock.

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 1 above)
Comment (niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c:111]: Thanks for the explanation. IMO, there always should be PU for CS pin, regardless of AP power state. But it is hardware change. Software change is good anyway.

File: subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c (hunk 2 above)
Comment (niedzwiecki-dawid) [reply subsys/mgmt/ec_host_cmd/backends/ec_host_cmd_backend_shi_ite.c]: Thanks, I hope I haven't missed anything.

**Review by niedzwiecki-dawid — APPROVED** [review]

**Review by GTLin08 — APPROVED** [review]

**Review by fabiobaltieri — APPROVED** [review]

**Pull request merged**